applogic/src/api/contacts_cubit.rs

```diff
@@ -0,0 +1,314 @@
+        let user_id: UserId = user_id
+            .map(From::from)
+            .unwrap_or_else(|| self.inner.own_user_id.clone());
+        let profile = self.inner.profiles.get(&user_id).cloned();
+        match profile {
+            Some(profile) => profile,
+            None => {
+                // spawn loading profile and return default profile
+                self.spawn_load_profile(user_id.clone());
+                let default_profile = UserProfile::from_user_id(&user_id);
+                UiUserProfile::from_profile(default_profile)
+            }
+        }
+    }
+
+    /// Returns the display name of the given user.
+    ///
+    /// If the user is not specificed, the display name of the logged-in user is returned.
+    ///
+    /// If the profile is not yet loaded, the default display name is returned and loading of the
+    /// profile is spawned in the background.
+    #[frb(sync)]
+    pub fn display_name(&self, user_id: Option<UiUserId>) -> String {
+        let user_id = user_id
+            .map(From::from)
+            .unwrap_or_else(|| self.inner.own_user_id.clone());
+        let display_name = self
+            .inner
+            .profiles
+            .get(&user_id)
+            .map(|profile| profile.display_name.clone());
+        match display_name {
+            Some(display_name) => display_name,
+            None => {
+                // spawn loading profile and return default display name
+                self.spawn_load_profile(user_id.clone());
+                DisplayName::from_user_id(&user_id).into_string()
+            }
+        }
+    }
+
+    /// Returns the profile picture of the given user if any is set.
+    ///
+    /// If the user is not specificed, the profile picture of the logged-in user is returned.
+    ///
+    /// If the profile is not yet loaded, `null` is returned and loading of the profile is spawned
+    /// in the background.
+    #[frb(sync)]
+    pub fn profile_picture(&self, user_id: Option<UiUserId>) -> Option<ImageData> {
+        let user_id = user_id
+            .map(From::from)
+            .unwrap_or_else(|| self.inner.own_user_id.clone());
+        let profile_picture = self
+            .inner
+            .profiles
+            .get(&user_id)
+            .map(|profile| profile.profile_picture.clone());
+        match profile_picture {
+            Some(profile_picture) => profile_picture,
+            None => {
+                // spawn loading profile
+                self.spawn_load_profile(user_id.clone());
+                None
+            }
+        }
+    }
+
+    #[frb(ignore)]
+    fn spawn_load_profile(&self, user_id: UserId) {
+        let tx = self.inner.load_profile_tx.clone();
+        spawn_from_sync(async move {
+            let _ = tx.send(user_id).await.inspect_err(|error| {
+                tracing::error!(%error, "Failed to send load profile request");
+            });
+        });
+    }
```

Comment:
I think (but I'm not sure) this could lead to multiple load requests at the same time. We might attempt to download the same profile several times.